feature: add ota updates for omi glass

kanishkatn opened the pull request:
# OTA Firmware Updates for Omi Glass

Note: Tested only on iOS.



## Changes

### Flutter
- Added firmware update method for BLE-based OTA
- Robust error handling and retry logic
- Code cleanup in firmware management

### OpenGlass Firmware
- Added dedicated OTA BLE service
- Implemented firmware transfer handling
- Fixed memory management issues during updates
- Added a release script 

## Testing

- Flash OpenGlass with firmware from this PR
- Create GitHub release with version 2.1.1. [Here's a sample](https://github.com/kanishkatn/omi/releases/tag/OpenGlass_v2.1.1)
- Test OTA update via Flutter app
- Verify successful firmware installation

## Demo

https://github.com/user-attachments/assets/341365ab-f08f-4c39-8774-4356ac178c38

## TODO

- test on Android and macOS

kanishkatn commented on backend/routers/firmware.py:
```diff
@@ -53,7 +53,7 @@ async def get_omi_github_releases(cache_key: str) -> Optional[List[Dict]]:
     # Make GitHub API request if not cached
     async with httpx.AsyncClient() as client:
-        url = "https://api.github.com/repos/BasedHardware/omi/releases?per_page=100"
+        url = "https://api.github.com/repos/kanishkatn/omi/releases?per_page=100"
```
```suggestion
        url = "https://api.github.com/repos/basedhardware/omi/releases?per_page=100"
```

beastoin commented:
@kanishkatn  cool!!! i guess it works well. btw, what is this? > the photo capture is currently disabled. haha. also, could you clean up your PR a bit:  

1/ the app already has the firmware download functionality - you don’t need to handle it, right?  

2/ please check the app > lib > services > devices > omi_connection.dart for having a better way to handle the custom OTA BLE services.  

3/ `const chunkSize = 480;` could we use the exchanged MTU to calculate the chunk size? some phones limit it to 247 or even 30 bytes data length.

4/ why do we need to change the partition?
```board_build.partitions = huge_app.csv```  
```board_build.partitions = min_spiffs.csv```  

5/ it’s omiglass, not openglass. also, remove the print logs etc.

aaravgarg commented:
@kanishkatn any updates?

kanishkatn commented:
Hello @beastoin @aaravgarg ,

I have addressed the following in the [commit](https://github.com/BasedHardware/omi/pull/2692/commits/3c8f3f6ecda696b22bd2e6c93273f6ab73fcdf2d) 

1. const chunkSize = 480; could we use the exchanged MTU to calculate the chunk size? some phones limit it to 247 or even 30 bytes data length. 

I've replaced the hardcoded chunkSize with a dynamic calculation. Here are the details

- Replaced hardcoded 480 bytes with MTU - 3 - 20 (ATT header + safety margin)
- Clamps between 20-500 bytes for compatibility with restrictive phones and ESP32 limits
- Accounts for 3-byte ATT header and 20-byte BLE stack safety margin

> 2/ please check the app > lib > services > devices > omi_connection.dart for having a better way to handle the custom OTA BLE services.

However, for the above feedback, I wasn't able to find a module/method that already handles OTA updates. The module used in this PR seems to be used for OTA updates on the omi necklace as well. 

If the feedback was to move the OTA updates completely to `omi_connection.dart`, I think it is out of scope for this PR. 

I have rebased the branch and have ensured that the OTA updates still work in my iPhone. Please let me know if you need any further inputs or assistance!

beastoin commented:
hey man @kanishkatn, Tu will help you finish since my room for the glass is near zero for now (Tu will talk to me in person so it would be more effective and help us move faster). 

this ticket, so let's close fn; then you guys will reopen it once everything is ready.

why i don't merge this PR? the friction is still large, and i think we should put more effort into making the implementation better. it would help the future maintainer a lot.

thank you.
